accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+object as we navigate the data structure. For instance, we need to
+create a wrapper `TinyJson(element)` in order to invoke `leaves`
+recursively.
+
+In contrast, the inline class is zero-cost, in the sense that it does
+_not_ use a wrapper object, it enforces the desired discipline
+statically. In the inline class, the invocation of `TinyJson(element)`
+in the body of `leaves` can be eliminated entirely by inlining.
+
+Inline classes are static in nature, like extension members: an inline
+class declaration may declare some type parameters. The type
+parameters will be bound to types which are determined by the static
+type of the receiver. Similarly, members of an inline type are resolved
+statically, i.e., if `tiny.leaves` is an invocation of an inline getter
+`leaves`, then the declaration named `leaves` whose body is executed
+is determined at compile-time. There is no support for late binding of
+an inline member, and hence there is no notion of overriding. In return
+for this lack of expressive power, we get improved performance.
+
+
+## Syntax
+
+A rule for `<inlineClassDeclaration>` is added to the grammar, along
+with some rules for elements used in inline class declarations:
+
+```ebnf
+<inlineClassDeclaration> ::=
+  'inline' 'class' <typeIdentifier> <typeParameters>? <interfaces>?
+  '{'
+    (<metadata> <inlineMemberDeclaration>)*
+  '}'
+
+<inlineMemberDeclaration> ::= <classMemberDefinition>
+```
+
+*The token `inline` is not made a built-in identifier: the reserved
+word `class` that occurs right after `inline` serves to disambiguate
+the inline class declaration with a fixed lookahead.*
+
+A few errors can be detected immediately from the syntax:
+
+A compile-time error occurs if the inline class does not declare any
+instance variables, and if it declares two or more instance
+variables. Let `v` be the name of unique instance variable that it
+declares. The declaration of `v` must have the modifier `final`, and it
+can not have the modifier `late`; otherwise a compile-time error
+occurs.
+
+The _name of the representation_ in an inline class declaration is the
+name `id` of the unique final instance variable that it declares, and
+the _type of the representation_ is the declared type of `id`.
+
+A compile-time error occurs if an inline class declaration declares an
+abstract member.
+
+*There are no special rules for static members in inline classes. They
+can be declared and called or torn off as usual, e.g.,
+`Inline.myStaticMethod(42)`.*
+
+
+## Primitives
+
+This document needs to refer to explicit inline method invocations, so we
+will add a special primitive, `invokeInlineMethod`, to denote invocations of
+inline methods.
+
+`invokeInlineMethod` is used as a specification device and it cannot occur in
+Dart source code. (*As a reminder of this fact, it uses syntax which is not
+derivable in the Dart grammar.*)
+
+
+### Static Analysis of invokeInlineMethod
```

Comment:
We may want to do that by introducing a generalization of extension methods and inline class methods, say "statically dispatched instance methods" (methods with a `this`, as opposed to, necessarily statically dispatched, static methods). The resolution of extension methods and inline class methods differ, but the invocation and tear-off behavior is the same (closes over actual receiver which `this` is bound to, at some type, plus type arguments provided at the invocation.)
Rather than saying that inline class methods work like extension methods, define how they both work, and say that both work like that, not like each other.

Reply:
> This doesn't cover tear offs, similarly for the dynamic semantics below.

Tear offs are covered in the new descriptions where `invokeInlineMethod` has been replaced by a natural language phrase.

> Perhaps it would be simpler to explain these directly

Did that (in terms of their semantics, no references to extension methods).

> define how they both work

This is the approach which is very nearly used already. This document specifies the semantics of tear offs of inline class methods directly, the language specification specifies the semantics of tear offs of extension methods directly, and then we just need to merge them when we integrate inline classes into the language specification.

I'll resolve this thread: I think there's no doubt that this issue has been resolved.